## Microbiome Analysis Training

Katie McCauley

[![Binder](https://mybinder.org/badge_logo.svg)](https://mybinder.org/v2/gh/ktmbiome-niaid/microbiome-analysis/new-interactive-approaches?urlpath=rstudio)

Date: TBD

Time: TBD

---

Thanks for your interest in the training! This GitHub repository contains the [lecture notes](https://htmlpreview.github.io/?https://github.com/ktmbiome-niaid/microbiome-analysis/training.html) and additional datasets and materials that will be used.

We have relatively minimal setup steps that are needed before you join us, which you can find on the [Preparations page](preparations.md).

I have also created a Binder for this training, which can be accessed through the badge. This is a workspace that has all packages pre-installed for you.

If you would like, you can process your own data on Nephele (https://nephele.niaid.nih.gov) and download the results zip to your computer. We'll focus on analyzing results generated from the DADA2 (Amplicon Metagenomics) or WGSA2 (Shotgun Metagenomics) pipelines.

This training was developed by the Bioinformatics and Computational Biosciences Branch at NIH/NIAID.
