## Microbiome Analysis Training

Katie McCauley

[![Binder](https://mybinder.org/badge_logo.svg)](https://mybinder.org/v2/gh/ktmbiome-niaid/microbiome-analysis/HEAD?urlpath=rstudio)

Date: Friday, November 8th, 2024

Time: 2-5 PM ET

---

Thank you for your interest in the training! This GitHub repository contains the [lecture notes](https://htmlpreview.github.io/?https://github.com/ktmbiome-niaid/microbiome-analysis/training.html) and additional datasets and materials that will be used. You can also [read the paper](infant_skin_manus2020.pdf) that our practice data is derived from.

We have relatively minimal setup steps that are needed before you join us, which you can find on the [Preparations page](preparations.md).

Alternatively, I have created a Binder for this training, which can be accessed through the badge above. This is an Rstudio workspace that you access through your browser and has all the packages we will be using for the training pre-installed.

If you would like, you can process your own data on [Nephele (nephele.niaid.nih.gov)](https://nephele.niaid.nih.gov) and download the results to your computer. We'll focus on analyzing results generated from the Amplicon Metagenomics (DADA2 or QIIME2) or Shotgun Metagenomics pipelines. If you have a dataset that is different from these two workflows, reach out in advance so I can advise whether this particular training will be applicable.

If you have any questions, reach out to Katie at [kathryn.mccauley@nih.gov](mailto:kathryn.mccauley@nih.gov).

This training was developed by the Bioinformatics and Computational Biosciences Branch at NIH/NIAID.
